Distinguish or and pipe

yoavrotems opened the pull request:
https://github.com/aquasecurity/linux-bench/issues/49 
The problem in those tests was that the engine didn't separated between the different uses of pipeline, and therefore tried to run non-exist commands.

yoavrotems committed:
Change TextToCommand  func

Changed the TextToCommand function to differentiate between two use cases of |.
The first case, pipe between different commands.
The second, or operand inside a command.

yoavrotems committed:
Update check_test.go

lizrice commented on check/check_test.go:
```diff
@@ -192,3 +193,70 @@ func TestGetFirstValidSubCheck(t *testing.T) {
+	testCases := []TestCase{
+		{
+			// Exist command
+			command:  "/bin/grep",
```
I'm not sure if we can rely on this always being present, for example this test fails on my Mac:
```
--- FAIL: TestIsShellCommand (0.01s)
    /Users/liz/go/src/github.com/aquasecurity/bench-common/check/check_test.go:270: case 0 expected for isShellCommand("/bin/grep") to return: true
FAIL
```
because 
```
$ ls /bin/grep                                                       ✔ 
ls: /bin/grep: No such file or directory
```

lizrice commented on check/check.go:
```diff
@@ -56,7 +56,7 @@ func (audit Audit) Execute(customConfig ...interface{}) (result string, errMessa
 	}
 
-	res, err := exec.Command("sh", "-c", string(audit)).CombinedOutput()
+	res, err := exec.Command("sh", "-c", strings.Replace(string(audit), "\\|", "|", -1)).CombinedOutput()
```
I think we already dealt with any \\| within `textToCommand()` so this change is unnecessary? 

But more importantly... we have just gone to the trouble of converting `audit` to `commands`... and then we go and just execute `audit` here. So I am not sure we need to do all this separation into commands in the first place. (This explains something I commented on elsewhere.) 

I've pushed a draft PR https://github.com/aquasecurity/bench-common/pull/89 to explore the idea of removing this `textToCommand()` function that I think might be altogether unnecessary

yoavrotems replied:
I think that the thought behind those procedures was to validate input before running everything as a shell command, the real question is this really saving overhead and maybe could be potentially more secured?

lizrice commented on check/check.go:
```diff
@@ -210,7 +210,23 @@ func textToCommand(s string) (cmds []*exec.Cmd) {
+	var newCP []string
+	joinNextPart := false
+
+	// Separate between pipeline and or, differentiate by | and \\|
```
```suggestion
	// Generally we split into commands separated by |
	// However, if the | was preceded by \\ then we need to undo that split, and replace the | (without the backslashes) 
```

I think this comment would be a bit clearer

lizrice commented on check/check_test.go:
```diff
@@ -192,3 +193,70 @@ func TestGetFirstValidSubCheck(t *testing.T) {
+	// For each test case, we want get the different commands separated by | and exclude the \\|
+	// which stand for the use of | as OR operator and not as a pipe.
+	testCases := []TestCase{
+		{
```
```suggestion
		{
			auditCommand:          "a b c",
			expectedSlicedCommand: []string{"a b c"},
		},
		{
			auditCommand:          "a b | c",
			expectedSlicedCommand: []string{"a b", "c"},
		},
		{
			auditCommand:          "a b \\| c",
			expectedSlicedCommand: []string{"a b | c"},
		},
		{
			auditCommand:          "a b \\|\\x\\| c",
			expectedSlicedCommand: []string{"a b |\\x| c"},
		},		
		{
```

Suggesting some additional easier-to-read tests that make it a bit clearer what we are expecting

lizrice commented on check/check.go:
```diff
@@ -210,7 +210,23 @@ func textToCommand(s string) (cmds []*exec.Cmd) {
+			joinNextPart = false
+		} else {
+			newCP = append(newCP, strings.TrimSuffix(v, "\\"))
+		}
```
```suggestion
		vv := strings.Trim(v, "\\")
		if joinNextPart {
			newCP[len(newCP)-1] += "|" + vv
			joinNextPart = false
		} else {
			newCP = append(newCP, vv)
		}
```

Makes no difference to the execution but storing the trimmed version of `v` in a local variable `vv` is a bit easier to read, I think. 

Generally speaking `strings.Join` is more efficient but I think the benefits for readability here are worth it, wdyt?

lizrice commented on check/check_test.go:
```diff
@@ -192,3 +193,70 @@ func TestGetFirstValidSubCheck(t *testing.T) {
+			// Test for mixed case of pipeline and or cases
+			auditCommand: "grep -E -i \"(\\\\v\\|\\\\r\\|\\\\m\\|\\\\s)\\|$(grep '^ID=' /etc/os-release | cut -d= -f2 | sed -e 's/\"//g'))\" /etc/motd",
+			expectedSlicedCommand: []string{
+				"grep -E -i \"(\\\\v|\\\\r|\\\\m|\\\\s)|$(grep '^ID=' /etc/os-release",
+				"cut -d= -f2",
+				"sed -e 's/\"//g'))\" /etc/motd"},
```
I think this test is correct in the sense that given this input to `testToCommand` I agree this is what we'd expect the output to be from that function. 

But, if the command gets split up like this, I don't see how it can actually run. Here's why:

The benchmark says run the following command: 
```
grep -E -i "(\\v|\\r|\\m|\\s|$(grep '^ID=' /etc/os-release | cut -d= -f2 | sed -e 's/"//g'))" /etc/motd
```

The $(...) part means we need to execute that part and use the results. So we could write the same command like this: 

```
$ RESULT = $(grep '^ID=' /etc/os-release | cut -d= -f2 | sed -e 's/"//g')
$ echo $RESULT
ubuntu
$ grep -E -i "(\\v|\\r|\\m|\\s|$RESULT)" /etc/motd
```
The way the commands are currently split up, the first one attempts to run 

```
grep -E -i \"(\\v|\\r|\\m|\\s)|$(grep '^ID=' /etc/os-release"
```
which results in 

```
$ grep -E -i \"(\\v|\\r|\\m|\\s)|$(grep '^ID=' /etc/os-release"
-bash: syntax error near unexpected token `('
```

What I eventually realised is that we do all this work to split up the commands _and then we just run the original audit command anyway!_ See my comment elsewhere, and PR https://github.com/aquasecurity/bench-common/pull/89

yoavrotems replied:
Yes, all of this is meant to check if the commands are valid (ignoring the parameters), the part that really gets checked afterward is cmd.Path, which will be just /bin/grep, /bin/cat, /bin/sed.
As you mentioned after this validation audit will execute as is.

lizrice reviewed (changes requested):
This has turned into a bit of a saga! See my comments and my other PR https://github.com/aquasecurity/bench-common/pull/89 and then let's discuss

yoavrotems committed:
Update check.go

Remove the command checks, and added a Reason for the different warnings situations.
(Like changes made in https://github.com/aquasecurity/kube-bench/pull/547/

yoavrotems committed:
Update check_test.go

Added Liz suggested tests from https://github.com/aquasecurity/bench-common/pull/89/files#diff-422a3cfa0a3cffd35c2836c5aff92172

yoavrotems committed:
Update util.go

Change the output to be more detailed when command doesn't exist or has non exist parameter and etc'

yoavrotems committed:
Update check.go

yoavrotems committed:
Update check.go

yoavrotems commented:
Hello :) 
Took your suggestions from https://github.com/aquasecurity/bench-common/pull/89/ 
And added the Warn reason functionality https://github.com/aquasecurity/kube-bench/pull/547 which will give a better description when a test fails.
here are the changes:
1. If a command doesn't exist in the system, the test will fail and in the remediation section we will see the next output
<img width="226" alt="command not found" src="https://user-images.githubusercontent.com/12915002/81210222-40bd1f00-8fda-11ea-8579-b93087dc1597.PNG">
 2. If you try use non exist command parameter the test will fail, and in the remediation section we will see the next output
<img width="472" alt="remediation when error with parameter" src="https://user-images.githubusercontent.com/12915002/81210754-02742f80-8fdb-11ea-946b-c5e7fc125707.PNG">
3. If you try run a test without any tests item test will fail, and in the remediation section we will see the next output 
<img width="201" alt="no tests items" src="https://user-images.githubusercontent.com/12915002/81210831-246db200-8fdb-11ea-9acd-a5b9fd6dad7e.PNG">
4. If you are using constrains and failed to find a correct sub check  test will warn you, and in the remediation section we will see the next output 
<img width="386" alt="Failed to find a vaild subcheck" src="https://user-images.githubusercontent.com/12915002/81212758-05245400-8fde-11ea-89c4-1cb1d2584976.PNG">
And Just a remainder:

> Any case of FAIL in a non-scored tests will be WARN

 and this applied to the above as well

yoavrotems committed:
Update check_test.go

Removed test about not scored check, because it used to be WARN, now its PASS when its good and WARN when its fails.
https://github.com/yoavAqua/bench-common/blame/bugfix-manual-and-not-scored-checks-result-with-warn-status/check/check_test.go#L41

lizrice commented on check/check.go:
```diff
@@ -164,8 +160,9 @@ func (c *Check) Run(definedConstraints map[string][]string) {
 		subCheck = getFirstValidSubCheck(c.SubChecks, definedConstraints)
 
 		if subCheck == nil {
+			c.Reason = "Failed to find a valid sub check, check your constrains "
```
```suggestion
			c.Reason = "Failed to find a valid sub check, check your constraints "
```

lizrice commented on check/check.go:
```diff
@@ -164,8 +160,9 @@ func (c *Check) Run(definedConstraints map[string][]string) {
+			c.Reason = "Failed to find a valid sub check, check your constrains "
 			c.State = WARN
-			glog.V(1).Info("Failed to find a valid sub check, check ", c.ID)
+			glog.V(1).Info("Failed to find a valid sub check, check your constrains")
```
```suggestion
			glog.V(1).Info("Failed to find a valid sub check, check your constraints")
```

lizrice commented on check/check.go:
```diff
@@ -138,12 +123,23 @@ type Group struct {
+	// Since this is an Scored check
+	// without tests return a 'WARN' to alert
+	// the user that this check needs attention
+	if c.Scored && len(strings.TrimSpace(c.Type)) == 0 && c.Tests == nil {
```
Don't we want to warn in this case even if it's not Scored?

lizrice replied:
I would be OK with adding this as a separate issue so that we can get this one merged quickly

yoavrotems replied:
Oh I see, yes I just thought that if it's not scored then it would still be warn and not fail but there will be a difference in the warn reason, I guess that the only changes between a scored and not scored tests will need to be in failed tests status, WARN instead of FAIL yes? 
Or maybe we want any other differences? 
In any case I think that open a different PR will be best so we could marge this one, and then add or change small stuff in different PRs :)

yoavrotems replied:
Currently removed the c.Scored parameter and not scored and scored tests will behave the same in this case.
I opened a PR https://github.com/aquasecurity/bench-common/issues/91 about the subject so we could understand the way we want to deal with not scored and scored tests

lizrice commented on check/check.go:
```diff
@@ -39,25 +38,10 @@ type Audit string
-
 	res, err := exec.Command("sh", "-c", string(audit)).CombinedOutput()
-
+	// Errors mean the audit command failed, but that might be what we expect
+	// for example, if we grep for something that is not found, there is a non-zero exit code
+	// But it is a problem if we can't find one of the audit commands to execute
```
```suggestion
	// Errors mean the audit command failed, but that might be what we expect
	// for example, if we grep for something that is not found, there is a non-zero exit code
	// It is a problem if we can't find one of the audit commands to execute, but we deal 
	// with this case in (c *Check) Run() 
```

lizrice commented on check/check.go:
```diff
@@ -176,6 +173,11 @@ func (c *Check) Run(definedConstraints map[string][]string) {
+		c.Reason = out
+		// Make output more readable
+		if (errmsgs == "exit status 127" || errmsgs == "exit status 1") && strings.HasSuffix(out, "not found\n") {
+			c.Reason = strings.Replace(c.Reason, "sh: 1:", "Command", -1)
```
Does this work in the case where the exit status is 127 and not 1, doesn't the string contain "sh: 127" in that case? I don't mind if we raise this as a separate issue so you can merge this PR fast

yoavrotems replied:
I think that should be a new PR I tried to find a quick fix for pretty printing the output, basically every output on my ubuntu machine contained the sh 1:  which I thought will be the case but as your commented here in your mac its just  sh: , there is also different output per Linux flavor I guess as we saw in the tests that my machine and your machine had different output.
We need to discuss if it something that we want to deal with because we don't have to

yoavrotems replied:
Ok I opened issue in the subject https://github.com/aquasecurity/bench-common/issues/90
currently I leave it unchanged.
I'm aware about the problem with the tests in mac, but here in Travis its pass, so until I will take changes in according to our decision in the above PR I will leave this part the same

lizrice reviewed (approved):
Hi @yoavrotems please can you take the suggestions for corrections to spelling / comments and then let's go ahead and merge. There are a couple of things I think we should consider further: 

- There is an issue running these tests on Mac (presumably because the shell returns different results). 
```
--- FAIL: TestRunAuditCommands (0.04s)
    check_test.go:249: Test 2: expected output sh: 1: anything: not found, got sh: anything: command not found
    check_test.go:249: Test 7: expected output sh: 1: anything: not found, got sh: anything: command not found
```

- I would have thought that if there are no test cases that should result in a WARN, even if the test is Not Scored? (check.go line 141)

yoavrotems committed:
Update check/check.go

Co-authored-by: Liz Rice <liz@lizrice.com>

yoavrotems committed:
Update check/check.go

Co-authored-by: Liz Rice <liz@lizrice.com>

yoavrotems committed:
Update check/check.go

Co-authored-by: Liz Rice <liz@lizrice.com>

yoavrotems committed:
Update check.go

mozillazg commented on check/check.go:
```diff
@@ -39,25 +38,11 @@ type Audit string
-		return result, errMessage, WARN
-	}
-
 	res, err := exec.Command("sh", "-c", string(audit)).CombinedOutput()
```
`sh -c` don't support multiple lines audit, how about:

```
cmd := exec.Command("sh")
cmd.Stdin = strings.NewReader(string(audit))
res, err := cmd.CombinedOutput()
```
